stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -24,54 +24,773 @@ internal func __NSDataInvokeDeallocatorFree(_ mem: UnsafeMutableRawPointer, _ le
 @_silgen_name("_NSWriteDataToFile_Swift")
 internal func _NSWriteDataToFile_Swift(url: NSURL, data: NSData, options: UInt, error: NSErrorPointer) -> Bool
 
-internal final class _SwiftNSData : _SwiftNativeNSData, _SwiftNativeFoundationType {
-    internal typealias ImmutableType = NSData
-    internal typealias MutableType = NSMutableData
+public final class _DataStorage {
+    public enum Backing {
+        // A mirror of the objective-c implementation that is suitable to inline in swift
+        case swift
+        
+        // these two storage points for immutable and mutable data are reserved for references that are returned by "known"
+        // cases from Foundation in which implement the backing of struct Data, these have signed up for the concept that
+        // the backing bytes/mutableBytes pointer does not change per call (unless mutated) as well as the length is ok
+        // to be cached, this means that as long as the backing reference is retained no further objc_msgSends need to be
+        // dynamically dispatched out to the reference.
+        case immutable(NSData) // This will most often (perhaps always) be NSConcreteData
+        case mutable(NSMutableData) // This will often (perhaps always) be NSConcreteMutableData
+        
+        // These are reserved for foregin sources where neither Swift nor Foundation are fully certain whom they belong
+        // to from a object inheritance standpoint, this means that all bets are off and the values of bytes, mutableBytes,
+        // and length cannot be cached. This also means that all methods are expected to dynamically dispatch out to the
+        // backing reference.
+        case customReference(NSData) // tracks data references that are only known to be immutable
+        case customMutableReference(NSMutableData) // tracks data references that are known to be mutable
+    }
+    
+    public static let maxSize = Int.max >> 1
+    public static let vmOpsThreshold = NSPageSize() * 4
+    
+    public static func allocate(_ size: Int, _ clear: Bool) -> UnsafeMutableRawPointer? {
+        if clear {
+            return calloc(1, size)
+        } else {
+            return malloc(size)
+        }
+    }
+    
+    
+    public static func move(_ dest_: UnsafeMutableRawPointer, _ source_: UnsafeRawPointer?, _ num_: Int) {
+        var dest = dest_
+        var source = source_
+        var num = num_
+        if _DataStorage.vmOpsThreshold <= num && ((unsafeBitCast(source, to: Int.self) | unsafeBitCast(dest, to: Int.self)) & (NSPageSize() - 1)) == 0 {
```

Comment:
I tried that - it didn't seem to work as I expected (I ran into a compiler crash that I was unable to diagnose)